staging/src/k8s.io/client-go/tools/cache/shared_informer.go

```diff
@@ -660,18 +736,30 @@ func (p *sharedProcessor) run(stopCh <-chan struct{}) {
 	func() {
 		p.listenersLock.RLock()
 		defer p.listenersLock.RUnlock()
-		for _, listener := range p.listeners {
+		for listener := range p.listeners {
 			p.wg.Start(listener.run)
 			p.wg.Start(listener.pop)
 		}
 		p.listenersStarted = true
 	}()
 	<-stopCh
-	p.listenersLock.RLock()
-	defer p.listenersLock.RUnlock()
-	for _, listener := range p.listeners {
-		close(listener.addCh) // Tell .pop() to stop. .pop() will tell .run() to stop
-	}
+
+	func() {
+		p.listenersLock.Lock()
```

Comment:
I'm assuming you moved that to its own function to unlock before the final wg wait?

Reply:
Yes, however it is important that the wg is not used after releasing the lock. Since the `wg` is only used when `listenersStarted` is true, it is not a problem. However, I think it is a bit sensitive.